# Battle of the Neighborhoods

## Business problem

Dubai is a very dynamic city with many different neighborhoods both in the old/historic part of the city, which existed before discovery of oil and subsequent boost of economy, and the new part of the city full of magnificent masterminds of world renown architects. A lot of international workers or so called expats are moving to Dubai every year, some bringing along their families and loved ones. And this is when the key challenge arises - where to rent/buy an apartment/house? 
Each person or a family will have a certain set of criteria for choosing a home, starting from availability of a supermarket in a close proximity, presence of playgrounds/parks and nearby schools/nurseries, to variety of restaurants and cafes around the corner. 

The business problem that I have chosen for the project is identification of the most suitable neighborhood based on criteria provided by the client, who could be a real estate agent trying to serve the client or the prospective tenants/owners themselves.


## Data to be used

The main component of data to be used for this project is the data provided by Foursquare API about the neighborhoods in Dubai along with venues of various categories. Neighborhoods will be considered as family friendly if for instance there are nurseries/schools, playgrounds or parks, grocery stores, gyms, recreational facilities located in them. 

## Preparation and cleaning of data

In [1]:
import pandas as pd # library for data analsysis
import requests

In [2]:
# import dataset from Dubai Government with geospatial data about neighborhoods

#use beautifulSoup for reading the xml
from bs4 import BeautifulSoup as Soup

with open('Community.kml') as data:
    kml_soup = Soup(data, 'lxml-xml') # Parse as XML

l = []

placeMarks = kml_soup.find_all('Placemark')  
for placeMark in placeMarks:
    d = {}
    #for each neighborhood take the first longitude and latitude coordinates
    coordinates = placeMark.find_all('coordinates')
    for coordinate in coordinates:
        coords = coordinate.text.split(",")
        d["Longitude"] =coords[0]
        d["Latitude"] =coords[1]
    
    # for each neighborhood take the english name
    ExtendedDatas = placeMark.find_all('ExtendedData')
    for ExtendedData in ExtendedDatas:
        simpleDatas = ExtendedData.find_all('SimpleData')
        for simpleData in simpleDatas:
            simpleDataAttrs  = dict(simpleData.attrs)
            if simpleDataAttrs[u'name'] == "CNAME_E":
                d["Neighborhood"] = simpleData.text
    l.append(d)
# create a dataframe based on the list generated from xml    
neighborhoods = pd.DataFrame(l) 
neighborhoods.head()     

Latitude         Longitude         Neighborhood
0  25.2651156252942  55.3011216866639           AL HAMRIYA
1  25.2595208149389   55.287777388341    AL SOUQ AL KABEER
2  25.1473131093492  55.2876875487575  NADD AL SHIBA FIRST
3  25.3119978356916   55.349689019013            AL MAMZAR
4  25.1884458878827  55.2410061879357        JUMEIRA THIRD

In [3]:
neighborhoods.shape

(226, 3)

In [4]:
neighborhoods

Latitude         Longitude                       Neighborhood
0    25.2651156252942  55.3011216866639                         AL HAMRIYA
1    25.2595208149389   55.287777388341                  AL SOUQ AL KABEER
2    25.1473131093492  55.2876875487575                NADD AL SHIBA FIRST
3    25.3119978356916   55.349689019013                          AL MAMZAR
4    25.1884458878827  55.2410061879357                      JUMEIRA THIRD
5    25.1473131070924  55.2876875487583  HADAEQ SHEIKH MOHAMMED BIN RASHID
6    25.0523737867498  55.4076823230994                    UMM NAHAD THIRD
7    24.8975228512717  54.9407412384163                      SAIH SHUAIB 1
8    25.0568630048816  55.4053403994602                    UMM NAHAD FIRST
9    25.0424225344569   55.314737341017                   UMM NAHAD SECOND
10   25.0219314544169  55.3251733771381                       AL YALAYIS 1
11   24.9987627956229  55.4028802168819                   UMM NAHAD FOURTH
12   25.1960139150478  55.2318250369513                     JUMEIRA SECOND
13   25.1831863435418   55.380157086628             RAS AL KHOR IND. THIRD
14   24.7222664107786  55.4011538721935                        AL O'SHOOSH
15   24.7328676863636  55.2028350521651                         AL LAYAN 2
16   25.2957254203803    55.38842208342                    AL NAHDA SECOND
17   25.0762137976922  55.5252902912887                         LE HEMAIRA
18   24.6918175490864  55.4012056868235                     SAIH SHUA'ALAH
19   25.2413157174406  55.3690331597278                         AL GARHOUD
20   25.1339472953113  55.2143813263652                          AL MANARA
21   25.2189142948569  55.3856386228249                        NADD SHAMMA
22   25.1365594958116  55.1853885804693                    AL SAFOUH FIRST
23   24.8618272128865  55.1955870530043                           MUGATRAH
24    25.088323856872  55.6049684078081                            ENKHALI
25   25.0827756592245  55.3351007382698                     WADI AL SAFA 4
26   25.2658045751317  55.2910881069959                       AL SHINDAGHA
27   25.2781664054421  55.3604856948222                      AL TWAR FIRST
28    25.011358166708   55.216198083957                     ME'AISEM FIRST
29   25.2369038497534   55.425852474489                    MUHAISNAH FIRST
..                ...               ...                                ...
196   25.268458438861  55.4005809602651                    AL QUSAIS THIRD
197  25.0006949110163  55.2274569920396                    ME'AISEM SECOND
198  25.2500568396936  55.3122685276605                          AL KARAMA
199  24.9236367035953  55.5212607840139                         AL LESAILY
200  25.0812609275417   55.249958347793              AL BARSHA SOUTH FIRST
201  25.1912901632221   55.408485488382                   AL WARQA'A FIRST
202  25.2984482828133  55.3967049363622               AL QUSAIS IND. THIRD
203  25.2575631934217  55.4559260709916                    AL MIZHAR FIRST
204  25.0047771847009  55.4657514633252                        AL YUFRAH 1
205  25.2175224786048  55.4392741193128                             MIRDIF
206  25.1402283171169  55.4655578612801                  AL ROWAIYAH FIRST
207  25.0636554023014  55.1986704000795              AL BARSHA SOUTH FIFTH
208  25.1667385858344  55.2321646006337                     AL SAFA SECOND
209  25.0343027146343  55.0225410504935                  NAKHLAT JABAL ALI
210  25.2740694878364  55.4610555727782               OUD AL MUTEENA THIRD
211  24.8005974512482  55.3827927310936                         AL HATHMAH
212  25.0792695052473   55.423206667635                       UMM AL DAMAN
213  25.1199009692512  55.3649438016522                     WADI AL SAFA 3
214  25.2745507853422  55.2938534850999         MADINAT DUBAI AL MELAHEYAH
215  25.2730376209521  55.1220840874846                      WORLD ISLANDS
216  25.1986142630199  55.2282445934063                   JUMEIRA ISLAND 2
217   25.254812801194  55.304266984844

so overall we have 226 neighborhoods in our dataset that we will use for analysis

convert the coordinates to float numbers

In [5]:
neighborhoods["Latitude"] = pd.to_numeric(neighborhoods["Latitude"])
neighborhoods["Longitude"] = pd.to_numeric(neighborhoods["Longitude"])

In [6]:
neighborhoods.dtypes

Latitude        float64
Longitude       float64
Neighborhood     object
dtype: object

Import all required librabries for further analysis

In [7]:
import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [8]:
address = 'Dubai, UAE'

geolocator = Nominatim(user_agent="tnt_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Dubai are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Dubai are 25.0750095, 55.1887608818332.


In [9]:
# create map of Dubai using latitude and longitude values
map_tnt = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Neighborhood']):
    label = '{}'.format(neighborhood)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tnt)  
    
map_tnt

setup client required for Foursquare

In [88]:
CLIENT_ID = '0BAJUB5WCRBTWR5P555BBPFQKHH5M4PVV4U5MQR5435QFIRH' # your Foursquare ID
CLIENT_SECRET = 'KTI0ETJNVRA0O4R4TCO1KZZKIBYSRMW1HXY3PLNO5YAFBTNW' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 20

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 0BAJUB5WCRBTWR5P555BBPFQKHH5M4PVV4U5MQR5435QFIRH
CLIENT_SECRET:KTI0ETJNVRA0O4R4TCO1KZZKIBYSRMW1HXY3PLNO5YAFBTNW


In [89]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [90]:
def getNearbyVenues(names, latitudes, longitudes, radius=400):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [94]:
dubai_venues = getNearbyVenues(names=neighborhoods['Neighborhood'],
                                   latitudes=neighborhoods['Latitude'],
                                   longitudes=neighborhoods['Longitude']
                                  )

AL HAMRIYA
AL SOUQ AL KABEER
NADD AL SHIBA FIRST
AL MAMZAR
JUMEIRA THIRD
HADAEQ SHEIKH MOHAMMED BIN RASHID
UMM NAHAD THIRD
SAIH SHUAIB 1
UMM NAHAD FIRST
UMM NAHAD SECOND
AL YALAYIS 1
UMM NAHAD FOURTH
JUMEIRA SECOND
RAS AL KHOR IND. THIRD
AL O'SHOOSH
AL LAYAN 2
AL NAHDA SECOND
LE HEMAIRA
SAIH SHUA'ALAH
AL GARHOUD
AL MANARA
NADD SHAMMA
AL SAFOUH FIRST
MUGATRAH
ENKHALI
WADI AL SAFA 4
AL SHINDAGHA
AL TWAR FIRST
ME'AISEM FIRST
MUHAISNAH FIRST
RAS AL KHOR
AL TWAR SECOND
AL HUDAIBA
LEHBAB FIRST
AL BARSHA THIRD
AL MUTEENA
UMM ESELAY
NAIF
TRADE CENTER FIRST
AL SATWA
LEHBAB SECOND
UMM HURAIR SECOND
BU KADRA
MANKHOOL
AL MIZHAR SECOND
WADI AL SAFA 6
AL MERYAL
UMM SUQEIM THIRD
AL THANYAH  FOURTH
NADD AL SHIBA THIRD
AYAL NASIR
OUD METHA
JABAL ALI FIRST
AL AWIR FIRST
HEFAIR
AL WARQA'A SECOND
AL QUSAIS IND. SECOND
OUD AL MUTEENA FIRST
WADI AL SAFA 5
AL JAFILIYA
JABAL ALI SECOND
AL HAMRIYA PORT
AL ROWAIYAH SECOND
AL YALAYIS 2
WARSAN SECOND
AL QUSAIS IND. FIRST
AL SAFOUH SECOND
HESSYAN SECOND
HOR AL ANZ

In [95]:
dubai_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   AL HAMRIYA              25.265116               55.301122   
1   AL HAMRIYA              25.265116               55.301122   
2   AL HAMRIYA              25.265116               55.301122   
3   AL HAMRIYA              25.265116               55.301122   
4   AL HAMRIYA              25.265116               55.301122   

                                               Venue  Venue Latitude  \
0  Sheikh Mohamed Centre For Cultural Understandi...       25.264091   
1                                      Coffee Museum       25.263633   
2                                     Baskin Robbins       25.264450   
3    Arabian Tea House Cafe (كافية بيت الشاي العربي)       25.263399   
4                                     Sikka Art Fair       25.264001   

   Venue Longitude  Venue Category  
0        55.300795   Historic Site  
1        55.300122          Museum  
2        55.299994  Ice Cream Shop  
3        55.299695        Tea Room  
4        55.300027     Art Gallery

In [96]:
print(dubai_venues.shape)

(1119, 7)


In [97]:
#Count the venues per each neighborhood
dubai_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                               
ABU HAIL                                               4   
AL BADA'                                              20   
AL BARAHA                                              4   
AL BARSHA FIRST                                       18   
AL BARSHA SECOND                                       7   
AL BARSHA SOUTH FIFTH                                  1   
AL BARSHA SOUTH FIRST                                  1   
AL BARSHA SOUTH THIRD                                  2   
AL BUTEEN                                             11   
AL CORNICHE                                            8   
AL DAGHAYA                                             9   
AL GARHOUD                                            20   
AL HAMRIYA                                            19   
AL HAMRIYA PORT                                        8   
AL HEBIAH FIRST                                        4   
AL HEBIAH FOURTH                                      11   
AL HUDAIBA                                             7   
AL JADAF                                               2   
AL KARAMA                                             16   
AL KHABAISI                                           13   
AL KHEERAN                                             4   
AL KHEERAN FIRST                                       5   
AL KHWANEEJ FIRST                                      1   
AL KHWANEEJ SECOND                                     1   
AL KIFAF                                              11   
AL MAMZAR                                              4   
AL MANARA                                             19   
AL MERKADH                                             2   
AL MIZHAR SECOND                                       4   
AL MURAQQABAT                                         20   
AL MURAR                                               8   
AL MUTEENA                                            20   
AL NAHDA FIRST                                        20   
AL NAHDA SECOND                                        4   
AL QOUZ FIRST                                          4   
AL QOUZ FOURTH                                         2   
AL QOUZ IND.FIRST                                      4   
AL QOUZ IND.FOURTH                                     4   
AL QOUZ IND.SECOND                                     5   
AL QOUZ IND.THIRD                                     16   
AL QOUZ THIRD                                          1   
AL QUSAIS FIRST                                        9   
AL QUSAIS IND. FIFTH                                   2   
AL QUSAIS IND. FIRST                                   5   
AL QUSAIS IND. FOURTH                                  2   
AL QUSAIS IND. SECOND                                  4   
AL QUSAIS IND. THIRD                                   4   
AL QUSAIS SECOND                                       4   
AL QUSAIS THIRD                                        4   
AL RAFFA                                              20   
AL RAS                                                10   
AL RASHIDIYA                                           4   
AL RIGGA                                              20   
AL SABKHA                                             20   
AL SAFA FIRST                                         20   
AL SAFA SECOND                                         2   
AL SAFOUH FIRST                                       20   
AL SAFOUH SECOND                                      20   
AL SATWA                                              13   
AL SHINDAGHA                                          17   
AL SOUQ AL KABEER                                     16   
AL THANYAH FIFTH                                       7   
AL THANYAH FIRST                                      20   
AL THANYAH THIRD                                      14   
AL TTAY                                                2   
AL TWAR FIRST  

In [98]:
print('There are {} uniques categories.'.format(len(dubai_venues['Venue Category'].unique())))

There are 200 uniques categories.


In [99]:
# one hot encoding
dubai_onehot = pd.get_dummies(dubai_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
dubai_onehot['Neighborhood_Name'] = dubai_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [dubai_onehot.columns[-1]] + list(dubai_onehot.columns[:-1])
dubai_onehot = dubai_onehot[fixed_columns]

dubai_onehot.head()

Neighborhood_Name  Accessories Store  Afghan Restaurant  African Restaurant  \
0        AL HAMRIYA                  0                  0                   0   
1        AL HAMRIYA                  0                  0                   0   
2        AL HAMRIYA                  0                  0                   0   
3        AL HAMRIYA                  0                  0                   0   
4        AL HAMRIYA                  0                  0                   0   

   Airport  Airport Food Court  Airport Lounge  American Restaurant  Arcade  \
0        0                   0               0                    0       0   
1        0                   0               0                    0       0   
2        0                   0               0                    0       0   
3        0                   0               0                    0       0   
4        0                   0               0                    0       0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0            0                    0                 0                   0   
1            0                    0                 0                   0   
2            0                    0                 0                   0   
3            0                    0                 0                   0   
4            1                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Badminton Court  Bakery  \
0            0              0          0           0                0       0   
1            0              0          0           0                0       0   
2            0              0          0           0                0       0   
3            0              0          0           0                0       0   
4            0              0          0           0                0       0   

   Bar  Basketball Court  Beach  Bed & Breakfast  Beer Garden  Bike Trail  \
0    0                 0      0                0            0           0   
1    0                 0      0                0            0           0   
2    0                 0      0                0            0           0   
3    0                 0      0                0            0           0   
4    0                 0      0                0            0           0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Breakfast Spot  Buffet  \
0       0              0          0         0               0       0   
1       0              0          0         0               0       0   
2       0              0          0         0               0       0   
3       0              0          0         0               0       0   
4       0              0          0         0               0       0   

   Burger Joint  Burrito Place  Bus Station  Business Service  Cafeteria  \
0             0              0            0                 0          0   
1             0              0            0                 0          0   
2             0              0            0                 0          0   
3             0              0            0                 0          0   
4             0              0            0                 0          0   

   Café  Campground  Canal  Carpet Store  Castle  Chinese Restaurant  \
0     0           0      0             0       0                   0   
1     0           0      0             0       0                   0   
2     0           0      0             0       0                   0   
3     0           0      0             0       0                   0   
4     0           0      0             0       0                   0   

   Chocolate Shop  Clothing Store  Cocktail Bar  Coffee Shop  Comedy Club  \
0               0               0             0            0            0   
1               0               0             0            0            0   
2               0               0             0            0       

In [100]:
# let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category
dubai_grouped = dubai_onehot.groupby('Neighborhood_Name').mean().reset_index()
dubai_grouped.head()

Neighborhood_Name  Accessories Store  Afghan Restaurant  African Restaurant  \
0          ABU HAIL                0.0           0.000000                 0.0   
1          AL BADA'                0.0           0.000000                 0.0   
2         AL BARAHA                0.0           0.000000                 0.0   
3   AL BARSHA FIRST                0.0           0.055556                 0.0   
4  AL BARSHA SECOND                0.0           0.000000                 0.0   

   Airport  Airport Food Court  Airport Lounge  American Restaurant  Arcade  \
0      0.0                 0.0             0.0                  0.0     0.0   
1      0.0                 0.0             0.0                  0.0     0.0   
2      0.0                 0.0             0.0                  0.0     0.0   
3      0.0                 0.0             0.0                  0.0     0.0   
4      0.0                 0.0             0.0                  0.0     0.0   

   Art Gallery  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0     0.000000             0.000000              0.00                 0.0   
1     0.000000             0.000000              0.15                 0.0   
2     0.000000             0.000000              0.00                 0.0   
3     0.000000             0.000000              0.00                 0.0   
4     0.142857             0.142857              0.00                 0.0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Badminton Court  Bakery  \
0     0.000000            0.0   0.000000         0.0              0.0    0.00   
1     0.000000            0.0   0.000000         0.0              0.0    0.00   
2     0.000000            0.0   0.000000         0.0              0.0    0.25   
3     0.055556            0.0   0.000000         0.0              0.0    0.00   
4     0.000000            0.0   0.142857         0.0              0.0    0.00   

   Bar  Basketball Court  Beach  Bed & Breakfast  Beer Garden  Bike Trail  \
0  0.0               0.0    0.0              0.0          0.0         0.0   
1  0.0               0.0    0.0              0.0          0.0         0.0   
2  0.0               0.0    0.0              0.0          0.0         0.0   
3  0.0               0.0    0.0              0.0          0.0         0.0   
4  0.0               0.0    0.0              0.0          0.0         0.0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Breakfast Spot  Buffet  \
0     0.0            0.0        0.0       0.0             0.0     0.0   
1     0.0            0.0        0.0       0.0             0.0     0.0   
2     0.0            0.0        0.0       0.0             0.0     0.0   
3     0.0            0.0        0.0       0.0             0.0     0.0   
4     0.0            0.0        0.0       0.0             0.0     0.0   

   Burger Joint  Burrito Place  Bus Station  Business Service  Cafeteria  \
0      0.000000            0.0          0.0               0.0       0.25   
1      0.000000            0.0          0.0               0.0       0.05   
2      0.000000            0.0          0.0               0.0       0.00   
3      0.055556            0.0          0.0               0.0       0.00   
4      0.000000            0.0          0.0               0.0       0.00   

       Café  Campground  Canal  Carpet Store  Castle  Chinese Restaurant  \
0  0.000000         0.0    0.0           0.0     0.0                0.00   
1  0.100000         0.0    0.0           0.0     0.0                0.05   
2  0.000000         0.0    0.0           0.0     0.0                0.00   
3  0.055556         0.0    0.0           0.0     0.0                0.00   
4  0.000000         0.0    0.0           0.0     0.0                0.00   

   Chocolate Shop  Clothing Store  Cocktail Bar  Coffee Shop  Comedy Club  \
0             0.0        0.000000           0.0     0.000000          0.0   
1             0.0        0.000000           0.0     0.050000          0.0   
2             0.0        0.000000          

In [101]:
#print each neighborhood along with the top 5 most common venues
num_top_venues = 5

for hood in dubai_grouped['Neighborhood_Name']:
    print("----"+hood+"----")
    temp = dubai_grouped[dubai_grouped['Neighborhood_Name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----ABU HAIL----
                  venue  freq
0           Pizza Place  0.25
1                  Park  0.25
2            Hookah Bar  0.25
3             Cafeteria  0.25
4  Other Great Outdoors  0.00


----AL BADA'----
                       venue  freq
0           Asian Restaurant  0.15
1                       Café  0.10
2  Middle Eastern Restaurant  0.10
3                Pizza Place  0.05
4                      Hotel  0.05


----AL BARAHA----
             venue  freq
0     Dessert Shop  0.25
1       Sports Bar  0.25
2             Park  0.25
3           Bakery  0.25
4  Organic Grocery  0.00


----AL BARSHA FIRST----
                 venue  freq
0          Pizza Place  0.11
1                  Gym  0.11
2       Clothing Store  0.06
3          Coffee Shop  0.06
4  Sporting Goods Shop  0.06


----AL BARSHA SECOND----
                       venue  freq
0                Art Gallery  0.14
1        Arts & Crafts Store  0.14
2                  BBQ Joint  0.14
3  Middle Eastern Restaurant  0.14
4 

In [103]:
#a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [134]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood_Name']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood_Name'] = dubai_grouped['Neighborhood_Name']

for ind in np.arange(dubai_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dubai_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood_Name 1st Most Common Venue      2nd Most Common Venue  \
0          ABU HAIL           Pizza Place                 Hookah Bar   
1          AL BADA'      Asian Restaurant  Middle Eastern Restaurant   
2         AL BARAHA          Dessert Shop                       Park   
3   AL BARSHA FIRST                   Gym                Pizza Place   
4  AL BARSHA SECOND        Ice Cream Shop                  BBQ Joint   

       3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                       Park             Cafeteria                  Farm   
1                       Café     Convenience Store             Cafeteria   
2                     Bakery            Sports Bar           Yoga Studio   
3          Indian Restaurant           Coffee Shop        Clothing Store   
4  Middle Eastern Restaurant           Art Gallery   Arts & Crafts Store   

  6th Most Common Venue    7th Most Common Venue 8th Most Common Venue  \
0            Food Court        Food & Drink Shop           Flower Shop   
1                 Hotel                Juice Bar   Fried Chicken Joint   
2  Fast Food Restaurant               Food Truck            Food Court   
3  Fast Food Restaurant                     Café          Burger Joint   
4  Gym / Fitness Center  Health & Beauty Service            Food Truck   

  9th Most Common Venue 10th Most Common Venue  
0           Flea Market          Fishing Store  
1  Pakistani Restaurant           Dessert Shop  
2     Food & Drink Shop            Flower Shop  
3        Sandwich Place             Food Truck  
4            Food Court      Food & Drink Shop

In [135]:
# Create clusters for the neghborhoods based on the nearby venues
# set number of clusters
kclusters = 8

dubai_grouped_clustering = dubai_grouped.drop('Neighborhood_Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dubai_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 3, 0, 0, 0, 0], dtype=int32)

In [136]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

dubai_merged = neighborhoods

# merge dubai_grouped with dubai_data to add latitude/longitude for each neighborhood
dubai_merged = dubai_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood_Name'), on='Neighborhood',how='inner')

dubai_merged.head() # check the last columns!

Latitude  Longitude         Neighborhood  Cluster Labels  \
0  25.265116  55.301122           AL HAMRIYA               0   
1  25.259521  55.287777    AL SOUQ AL KABEER               0   
2  25.147313  55.287688  NADD AL SHIBA FIRST               0   
3  25.311998  55.349689            AL MAMZAR               0   
4  25.188446  55.241006        JUMEIRA THIRD               0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Art Gallery                  Café         Historic Site   
1                 Hotel             Nightclub    Persian Restaurant   
2            Bike Trail           IT Services                   Spa   
3                 Beach              Tea Room   Fried Chicken Joint   
4                  Café           Coffee Shop        Ice Cream Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                 Hotel           Coffee Shop                Museum   
1      Asian Restaurant             Roof Deck             Gastropub   
2  Gym / Fitness Center        Farmers Market            Food Truck   
3            Food Truck            Food Court     Food & Drink Shop   
4  Gym / Fitness Center      Department Store                 Plaza   

  7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0              Tea Room  Middle Eastern Restaurant         Boat or Ferry   
1     Electronics Store          Korean Restaurant    Chinese Restaurant   
2            Food Court          Food & Drink Shop           Flower Shop   
3           Flower Shop                Flea Market         Fishing Store   
4            Donut Shop               Burger Joint     French Restaurant   

  10th Most Common Venue  
0             Donut Shop  
1            Supermarket  
2            Flea Market  
3           Fishing Spot  
4          Grocery Store

In [137]:
dubai_merged.dtypes

Latitude                  float64
Longitude                 float64
Neighborhood               object
Cluster Labels              int32
1st Most Common Venue      object
2nd Most Common Venue      object
3rd Most Common Venue      object
4th Most Common Venue      object
5th Most Common Venue      object
6th Most Common Venue      object
7th Most Common Venue      object
8th Most Common Venue      object
9th Most Common Venue      object
10th Most Common Venue     object
dtype: object

In [138]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dubai_merged['Latitude'], dubai_merged['Longitude'], dubai_merged['Neighborhood'], dubai_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[(cluster)-1],
        fill=True,
        fill_color=rainbow[(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [147]:
#Go through each of the clusters to identify the basis for grouping them together
#CLUSTER 0
dubai_merged.loc[dubai_merged['Cluster Labels'] == 0, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of very diverse neighborhoods, which could be used by the real estate agents to filter based on customer preferences

Neighborhood      1st Most Common Venue  \
0                           AL HAMRIYA                Art Gallery   
1                    AL SOUQ AL KABEER                      Hotel   
2                  NADD AL SHIBA FIRST                 Bike Trail   
3                            AL MAMZAR                      Beach   
4                        JUMEIRA THIRD                       Café   
5    HADAEQ SHEIKH MOHAMMED BIN RASHID                 Bike Trail   
7                        SAIH SHUAIB 1    Comfort Food Restaurant   
12                      JUMEIRA SECOND                      Beach   
13              RAS AL KHOR IND. THIRD           Department Store   
16                     AL NAHDA SECOND          Indian Restaurant   
19                          AL GARHOUD             Airport Lounge   
20                           AL MANARA                      Hotel   
21                         NADD SHAMMA                 Shoe Store   
22                     AL SAFOUH FIRST                        Bar   
26                        AL SHINDAGHA  Middle Eastern Restaurant   
27                       AL TWAR FIRST          Indian Restaurant   
29                     MUHAISNAH FIRST           Indoor Play Area   
30                         RAS AL KHOR                       Pier   
31                      AL TWAR SECOND                 Restaurant   
32                          AL HUDAIBA        Fried Chicken Joint   
35                          AL MUTEENA                      Hotel   
37                                NAIF                      Hotel   
38                  TRADE CENTER FIRST                 Restaurant   
39                            AL SATWA              Grocery Store   
41                   UMM HURAIR SECOND               Fishing Spot   
43                            MANKHOOL                Coffee Shop   
44                    AL MIZHAR SECOND  Middle Eastern Restaurant   
47                    UMM SUQEIM THIRD                      Hotel   
49                 NADD AL SHIBA THIRD                  Cafeteria   
50                          AYAL NASIR             Clothing Store   
51                           OUD METHA                       Café   
52                     JABAL ALI FIRST                       Pool   
55                   AL WARQA'A SECOND               Intersection   
56               AL QUSAIS IND. SECOND              Moving Target   
60                    JABAL ALI SECOND                       Pool   
61                     AL HAMRIYA PORT                  Cafeteria   
63                        AL YALAYIS 2         Athletics & Sports   
64                       WARSAN SECOND             Ice Cream Shop   
65                AL QUSAIS IND. FIRST              Auto Workshop   
66                    AL SAFOUH SECOND                      Hotel   
68                     HOR AL ANZ EAST  Middle Eastern Restaurant   
69                         AL CORNICHE                Pizza Place   
76                 TRADE CENTER SECOND                 Food Court   
79                    MUHAISANAH THIRD             Clothing Store   
82                    AL KHEERAN FIRST             Ice Cream Shop   
83                      AL QOUZ FOURTH               Intersection   
85                       AL MURAQQABAT  Middle Eastern Restaurant   
89                       NADD AL HAMAR  Middle Eastern Restaurant   
91                    AL BARSHA SECOND             Ice Cream Shop   
93                         AL KHABAISI                      Hotel   
96                            AL MURAR                      Hotel   
97                       AL TWAR THIRD          Convenience Store   
100                  AL QOUZ IND.FIRST         Photography Studio   
103                         PORT SAEED                       Café   
104         JABAL ALI INDUSTRIAL FIRST                Bus Station   
106            RAS AL KHOR IND. SECOND         African Restaurant   
107                   AL QUSAIS SECOND  Indian Chinese Restaurant   
112                     

In [139]:
#CLUSTER 1
dubai_merged.loc[dubai_merged['Cluster Labels'] == 1, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of neighborhoods suitable for lovers of yoga, camping, fishing, food trucks and flowers

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
11    UMM NAHAD FOURTH            Campground           Yoga Studio   
70    AL WARQA'A FIFTH          Intersection            Campground   
134  AL KHWANEEJ FIRST            Campground           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11      French Restaurant            Food Truck            Food Court   
70            Yoga Studio        Farmers Market            Food Truck   
134     French Restaurant            Food Truck            Food Court   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11      Food & Drink Shop           Flower Shop           Flea Market   
70             Food Court     Food & Drink Shop           Flower Shop   
134     Food & Drink Shop           Flower Shop           Flea Market   

    9th Most Common Venue 10th Most Common Venue  
11          Fishing Store           Fishing Spot  
70            Flea Market          Fishing Store  
134         Fishing Store           Fishing Spot

In [140]:
#CLUSTER 2
dubai_merged.loc[dubai_merged['Cluster Labels'] == 2, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of only one neighborhood with pet store and yoga studio as most common venue

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
194  NADD AL SHIBA FOURTH             Pet Store           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
194                  Farm            Food Truck            Food Court   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
194     Food & Drink Shop           Flower Shop           Flea Market   

    9th Most Common Venue 10th Most Common Venue  
194         Fishing Store           Fishing Spot

In [141]:
#CLUSTER 3
dubai_merged.loc[dubai_merged['Cluster Labels'] == 3, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of neighborhoods suitable for lovers of pools, yoga, farm life and food courts

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
45          WADI AL SAFA 6                  Pool           Yoga Studio   
207  AL BARSHA SOUTH FIFTH                  Pool           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45                   Farm            Food Truck            Food Court   
207                  Farm            Food Truck            Food Court   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45      Food & Drink Shop           Flower Shop           Flea Market   
207     Food & Drink Shop           Flower Shop           Flea Market   

    9th Most Common Venue 10th Most Common Venue  
45          Fishing Store           Fishing Spot  
207         Fishing Store           Fishing Spot

In [142]:
#CLUSTER 4
dubai_merged.loc[dubai_merged['Cluster Labels'] == 4, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists only one neighborhood suitable for barbeques and primarily restaurants/food courts as common venues

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
205       MIRDIF             BBQ Joint           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
205  Fast Food Restaurant     French Restaurant            Food Truck   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
205            Food Court     Food & Drink Shop           Flower Shop   

    9th Most Common Venue 10th Most Common Venue  
205           Flea Market          Fishing Store

In [143]:
#CLUSTER 5
dubai_merged.loc[dubai_merged['Cluster Labels'] == 5, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of neighborhoods suitable for lovers of mountains, indian cuisine, flea market and food courts

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
153               AL TTAY     Indian Restaurant              Mountain   
156  AL QUSAIS IND. FIFTH              Mountain        Shop & Service   
173         AL QOUZ THIRD     Indian Restaurant        Farmers Market   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
153          Dance Studio  Fast Food Restaurant     French Restaurant   
156        Farmers Market            Food Truck            Food Court   
173     French Restaurant            Food Truck            Food Court   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
153            Food Truck            Food Court     Food & Drink Shop   
156     Food & Drink Shop           Flower Shop           Flea Market   
173     Food & Drink Shop           Flower Shop           Flea Market   

    9th Most Common Venue 10th Most Common Venue  
153           Flower Shop            Flea Market  
156         Fishing Store           Fishing Spot  
173         Fishing Store           Fishing Spot

In [144]:
#CLUSTER 6
dubai_merged.loc[dubai_merged['Cluster Labels'] == 6, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of neighborhoods mainly with cafes, yoga studios, fast food restaurants and fishing stores

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
42                BU KADRA                  Café           Yoga Studio   
108        AL HEBIAH FIRST                  Café   Housing Development   
155  AL QUSAIS IND. FOURTH           Coffee Shop                  Café   
172   NADD AL SHIBA SECOND                  Café           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
42   Fast Food Restaurant     French Restaurant            Food Truck   
108                   Gym           Yoga Studio  Fast Food Restaurant   
155           Yoga Studio        Farmers Market            Food Truck   
172  Fast Food Restaurant     French Restaurant            Food Truck   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
42             Food Court     Food & Drink Shop           Flower Shop   
108            Food Truck            Food Court     Food & Drink Shop   
155            Food Court     Food & Drink Shop           Flower Shop   
172            Food Court     Food & Drink Shop           Flower Shop   

    9th Most Common Venue 10th Most Common Venue  
42            Flea Market          Fishing Store  
108           Flower Shop            Flea Market  
155           Flea Market          Fishing Store  
172           Flea Market          Fishing Store

In [148]:
#CLUSTER 7
dubai_merged.loc[dubai_merged['Cluster Labels'] == 7, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]

# this cluster consists of only one neighborhood with rental car location and yoga studios as most common venues

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
161   UMM RAMOOL   Rental Car Location           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
161                  Farm            Food Truck            Food Court   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
161     Food & Drink Shop           Flower Shop           Flea Market   

    9th Most Common Venue 10th Most Common Venue  
161         Fishing Store           Fishing Spot

In [184]:
def filt(df, cols, venues):
    df_result = df.loc[(df[cols[0]].isin(venues))]
    columns = cols[1:]
    for col1 in columns:
        df_small = df.loc[(df[col1].isin(venues))]
        df_result = df_result.append(df_small, ignore_index=True)
    
    return df_result

In [185]:
# define criteria for search
CLUSTER = 0
VENUES = ['Playground','Indoor Play Area','Gym','Yoga Studio','Pool','Park']
COLS = ["1st Most Common Venue","2nd Most Common Venue","3rd Most Common Venue","4th Most Common Venue","5th Most Common Venue","6th Most Common Venue","7th Most Common Venue","8th Most Common Venue","9th Most Common Venue","10th Most Common Venue"]

dubai_cluster = dubai_merged.loc[dubai_merged['Cluster Labels'] == CLUSTER, dubai_merged.columns[[2] + list(range(4, dubai_merged.shape[1]))]]
filtered_cluster = filt(dubai_cluster,COLS,VENUES )

In [186]:
filtered_cluster

Neighborhood      1st Most Common Venue  \
0              MUHAISNAH FIRST           Indoor Play Area   
1              JABAL ALI FIRST                       Pool   
2             JABAL ALI SECOND                       Pool   
3              AL BARSHA FIRST                        Gym   
4                SAIH SHUAIB 1    Comfort Food Restaurant   
5          NADD AL SHIBA THIRD                  Cafeteria   
6             MUHAISANAH THIRD             Clothing Store   
7           AL KHWANEEJ SECOND          Health Food Store   
8             AL HEBIAH FOURTH               Soccer Field   
9              ZAA'BEEL SECOND         Punjabi Restaurant   
10                   AL BARAHA               Dessert Shop   
11          AL QOUZ IND.FOURTH               Soccer Field   
12         DUBAI INT'L AIRPORT               Soccer Field   
13                UMM AL SHEIF              Boat or Ferry   
14                BURJ KHALIFA                      Hotel   
15               WARSAN FOURTH           Asian Restaurant   
16             AL QUSAIS THIRD                     Market   
17       AL BARSHA SOUTH FIRST                Art Gallery   
18             MUHAISNAH FIRST           Indoor Play Area   
19              AL QOUZ FOURTH               Intersection   
20               AL TWAR THIRD          Convenience Store   
21            UMM SUQEIM FIRST            Harbor / Marina   
22            UMM SUQEIM FIRST            Harbor / Marina   
23            UMM SUQEIM FIRST            Harbor / Marina   
24            UMM SUQEIM FIRST            Harbor / Marina   
25            UMM SUQEIM FIRST            Harbor / Marina   
26                  AL WUHEIDA               Soccer Field   
27          AL QOUZ IND.FOURTH               Soccer Field   
28                    AL KIFAF  Middle Eastern Restaurant   
29       AL BARSHA SOUTH THIRD                   Pharmacy   
30                    ABU HAIL                Pizza Place   
31           MUHAISANAH FOURTH          Currency Exchange   
32              WADI AL SAFA 2                Pizza Place   
33              AL SAFA SECOND         Seafood Restaurant   
34              WADI AL SAFA 3                Pizza Place   
35            JUMEIRA ISLAND 2                      Beach   
36                  AL MERKADH                  Gastropub   
37              JUMEIRA SECOND                      Beach   
38                   AL MANARA                      Hotel   
39              AL TWAR SECOND                 Restaurant   
40                AL YALAYIS 2         Athletics & Sports   
41            MUHAISANAH THIRD             Clothing Store   
42               NADD AL HAMAR  Middle Eastern Restaurant   
43                BUSINESS BAY                  Nightclub   
44         DUBAI INT'L AIRPORT               Soccer Field   
45  MADINAT DUBAI AL MELAHEYAH                        Bar   
46                  AL KHEERAN              Jewelry Store   
47                 NADD SHAMMA                 Shoe Store   
48           UMM HURAIR SECOND               Fishing Spot   
49            AL MIZHAR SECOND  Middle Eastern Restaurant   
50             JABAL ALI FIRST                       Pool   
51       AL QUSAIS IND. SECOND              Moving Target   
52            JABAL ALI SECOND                       Pool   
53        AL QUSAIS IND. FIRST              Auto Workshop   
54               NADD AL HAMAR  Middle Eastern Restaurant   
55               AL TWAR THIRD          Convenience Store   
56           AL QOUZ IND.FIRST         Photography Studio   
57                      NAZWAH                Supermarket   
58                AL RASHIDIYA                Bus Station   
59                   AL BARAHA               Dessert Shop   
60               AL QOUZ FIRST                      Canal   
61             AL QUSAIS THIRD                     Market   
62           AL WARQA'A SECOND               Intersection   
63     RAS AL KHOR IND. SECOND         African Restaurant   
64              CORNICHE DEIRA         Seafood Restau

## as a results, we have identified 78 neighborhoods that have venues matching the criteria given by the client